GoogleForCreators/web-stories-wp
# Right Click Menu: Implementation Proposal

## Context

Right click menu

## Summary

Initial idea for right click menu.

## Relevant Technical Choices

We'll be hooking into a lot of existing functionality, so I decided


This follows the [quick actions menu](https://github.com/google/web-stories-wp/blob/main/assets/src/edit-story/app/highlights/quickActions/useQuickActions.js) architecture pretty closely.

Right click menus will need the following information to determine what items to show and what the items do:
- the selected element (element that was right-clicked)
- story context: to know what styles will be copied/the element information. May also need this for determining the front/back placement of elements in relation to each other.
- page context: adding, removing, and duplicating pages

These items will hook into existing functionality most of the time. For that reason, I decided to follow the quick actions setup since it hooks into other contexts like the 'highlights' context.

## To-do

A few questions that need to be talked about:


- [x] Does right clicking an element select it? yes it does
- [ ] (other pr)Investigate and possibly refactor the way the context menu opens. If we can leverage using the `blur` event then we will be able to click something under the 'mask' of the context menu 


## User-facing changes

![right-click-menu](https://user-images.githubusercontent.com/22185279/121592145-8c20d880-c9f7-11eb-85e8-c64b78e6ba07.gif)

## Testing Instructions

1. Turn on the right click menu experiment
2. Go to editor
3. Add an element
4. Right right clicking the canvas should open the right click menu
5. Right clicking the element in the canvas should open the right click menu and select the element

**Note:** Right now you have to click away from the context menu to close it first before right clicking again. This will be fixed in a later ticket.

### QA






## Reviews

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist




---



Fixes #6156
